swim.go

```diff
@@ -279,35 +323,217 @@ func (s *SWIM) startFailureDetector() {
 // 3. At the end of T, SWIM checks to see if ack was received from k members, and if there is no message,
 //    The member(j) is judged to be failed, so check the member(j) as suspected or delete the member(j) from memberMap.
 //
-
-func (s *SWIM) probe(member Member) {
+func (s *SWIM) probe(member Member, timer *time.Timer) {
 
 	if member.Status == Dead {
 		return
 	}
 
-	end := make(chan struct{}, 1)
-	defer close(end)
+	end := make(chan ProbeResponse, 1)
+	defer func() {
+		close(end)
+	}()
+
+	// this context and wait group are used to cancel probe procedure
+	// when probing time is out
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	ctx, cancel := context.WithCancel(context.Background())
 
 	go func() {
+		task := func() (interface{}, error) {
+			err := s.ping(&member)
+			if err == nil {
+				return nil, nil
+			}
 
-		// Ping to member
-		time.Sleep(1 * time.Second)
-		end <- struct{}{}
-	}()
+			if err != ErrSendTimeout {
+				return nil, ErrPingFailed
+			}
+
+			err = s.indirectProbe(&member)
+			if err != nil {
+				return nil, err
+			}
 
-	T := time.NewTimer(time.Millisecond * time.Duration(s.config.T))
+			return nil, nil
+		}
+		resp := NewTaskRunner(task, ctx).Start()
+		end <- ProbeResponse{err: resp.err}
+	}()
```

Comment:
probing producer run with `TaskRunner` as when probing time is out, this goroutine should be canceled